File: packages/jexlang-ts/src/eval/scopes/ScopeStack.ts
import type { JexValue } from "../../types";

/**
 * Represents a single scope of local variables.
 */
export class Scope {
  private variables: Map<string, JexValue> = new Map();

  constructor() {}

  /**
   * Sets a variable in the current scope
   */
  set(name: string, value: JexValue): void {
    this.variables.set(name, value);
  }

  /**
   * Gets a variable from the current scope
   */
  get(name: string): JexValue | undefined {
    return this.variables.get(name);
  }

  /**
   * Checks if a variable exists in the current scope
   */
  has(name: string): boolean {
    return this.variables.has(name);
  }
}

/**
 * Manages a stack of variable scopes for handling local variables.
 */
export class ScopeStack {
  private scopes: Scope[] = [new Scope()]; // Start with global scope

  /**
   * Pushes a new scope onto the stack
   */
  pushScope(): Scope {
    const newScope = new Scope();
    this.scopes.push(newScope);
    return newScope;
  }

  /**
   * Pops the current scope off the stack
   */
  popScope(): Scope | undefined {
    if (this.scopes.length <= 1) return undefined; // Keep at least one scope
    return this.scopes.pop();
  }

  /**
   * Gets the current scope
   */
  currentScope(): Scope {
    return this.scopes[this.scopes.length - 1];
  }

  /**
   * Sets a variable in the current scope
   */
  set(name: string, value: JexValue): void {
    this.currentScope().set(name, value);
  }

  /**
   * Gets a variable from any scope, searching from most local to most global
   */
  get(name: string): JexValue | undefined {
    // Search scopes from most local to most global
    for (let i = this.scopes.length - 1; i >= 0; i--) {
      const scope = this.scopes[i];
      if (scope.has(name)) {
        return scope.get(name);
      }
    }
    return undefined;
  }

  /**
   * Checks if a variable exists in any scope
   */
  has(name: string): boolean {
    for (let i = this.scopes.length - 1; i >= 0; i--) {
      if (this.scopes[i].has(name)) {
        return true;
      }
    }
    return false;
  }
}

File: packages/jexlang-java/src/main/java/com/jexlang/java/scopes/ScopeStack.java
package com.jexlang.java.scopes;

import com.jexlang.java.types.JexValue;

import java.util.ArrayList;
import java.util.Collections;

public class ScopeStack {
    private ArrayList<Scope> scopes = new ArrayList<>(Collections.singletonList(new Scope())); // Start with global scope

    public Scope pushScope() {
        Scope scope = new Scope();
        scopes.add(scope);
        return scope;
    }

    public Scope popScope() {
        if (this.scopes.size() <= 1) {
            return null; // Keep at least one scope
        }
        return scopes.remove(scopes.size() - 1);
    }

    public Scope currentScope() {
        if (scopes.isEmpty()) {
            throw new RuntimeException("No current scope");
        }
        return scopes.get(scopes.size() - 1);
    }

    public void set(String name, JexValue value) {
        currentScope().set(name, value);
    }

    public JexValue get(String name) {
        for (int i = scopes.size() - 1; i >= 0; i--) {
            Scope scope = scopes.get(i);
            if (scope.has(name)) {
                return scope.get(name);
            }
        }
        return null;
    }

    public boolean has(String name) {
        for (int i = scopes.size() - 1; i >= 0; i--) {
            Scope scope = scopes.get(i);
            if (scope.has(name)) {
                return true;
            }
        }
        return false;
    }
}
